resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala

```diff
@@ -192,6 +193,21 @@ private[spark] class BasicExecutorFeatureStep(
           .endResources()
         .build()
     }.getOrElse(executorContainer)
+    val containerWithLifecycle = kubernetesConf.workerDecommissioning match {
+      case false =>
+        logInfo("Decommissioning not enabled, skipping shutdown script")
+        containerWithLimitCores
+      case true =>
+        logInfo("Adding decommission script to lifecycle")
+        new ContainerBuilder(containerWithLimitCores).withNewLifecycle()
+          .withNewPreStop()
```

Comment:
Will this get triggered when Spark itself stops the executor (i.e. when you turn on dynamic allocation)? Does the code behave as it should in that case?

Reply:
I'm not sure why it would not, all the other properties presumably work. Am I missing some special handling logic we have seperate for the pods for dynamic alloc?

Reply:
My question is more in the lines of: since this will be triggered by Spark itself killing executors, will everything else triggered by this script behave correctly when Spark itself, and not the cluster, is requesting that the executor be killed?

e.g. now you have an executor being killed that some other code might try to decommission and those are all tracked in separate places, and might get out of sync.

Reply:
Oh I think I see where this is going. So (for right now) when Spark is doing it's own internal down scaling we don't take advantage of that knowledge. We totally should in the future, that's actually my plan in part 5 but I wanted to keep the scope as small as possible for the first change.

Reply:
I was thinking more in terms of correctness; e.g. if the executor exits by itself (which is what dynamic allocation does), will the pre-exit script cause any errors in log files or the status of the container, things long those lines.

I think it's easier if you just try it (with {{spark.dynamicAllocation.enabled=true}} and {{spark.dynamicAllocation.shuffleTracking.enabled=true}}).

Reply:
So I'm not seeing how this might be an issue and I want to make sure I understand. Whats the exact situation where you think it would go bad/why?

Reply:
I'm asking you to make sure nothing can go bad. You're adding hooks in the shutdown path for when the cluster is shutting things down. With dynamic allocation, Spark it shutting things down itself. I want to ensure that you accounted for that, and that your solution doesn't cause problems in that scenario.

Reply:
Just to make sure I'm understanding the desire you want me to spin up a cluster with both these enabled, or change/add integration test to enable & trigger dynamic alloc and decom, or something else?

Reply:
Whatever you need to convince yourself that nothing bad is happening.

Reply:
So I did some looking in the code and how we remove executors during dynamic allocation is a different mechanism. When we kill executors in the K8s backend we first ask Spark to exit. If it doesn't exit (see L153) we do a delete. Now the decommissioning would be triggered in that event and could keep the executor alive for the length of the timeout period configured. I think that's ok given that the timeout period should be set something not too long, and most of the executors likely exit when asked by Spark and don't have to be cleaned up by K8s. That being said if that is not OK from https://kubernetes.io/docs/tasks/run-application/force-delete-stateful-set-pod/ it looks like K8s in 1.5+ supports force delition with no grace period regardless of the pods configuration.

Reply:
So when the Spark executor exits by itself, the pre-exit script is not executed?

Reply:
Indeed, the Kubernetes docs were updated to clarify how prestop hooks work https://github.com/kubernetes/website/pull/12498/files which indicates they only trigger on pod deletion, not when the containers main process exits gracefully.

Reply:
Cool, that answers my initial question and takes care of my concerns.